File: dart_cli/lib/structs/KeysTable.dart

import "../ByteDataWrapper.dart";
import "../ioClass.dart";
import "KeyMapping.dart";

class KeysTable extends IoClass {
  int count = 0;
  late final List<KeyMapping> mappings;
  
  KeysTable() :
    mappings = [];
  
  KeysTable.readBytes(ByteDataWrapper bytes) {
    count = bytes.readUint16();
    mappings = List.generate(count, (index) => KeyMapping.readBytes(bytes));
  }

  @override
  void writeBytes(ByteDataWrapper bytes) {
    bytes.writeUint16(count);
    for (var mapping in mappings)
      mapping.writeBytes(bytes);
  }

  @override
  int get size {
    var size = 2;
    for (var mapping in mappings)
      size += mapping.size;
    return size;
  }

  String lookupIndex(int index) {
    return mappings[index].keyName;
  }

  int? getKeyIndex(String key, {bool allowCreate = false}) {
    for (var i = 0; i < mappings.length; i++) {
      if (mappings[i].keyName == key)
        return i;
    }
    if (!allowCreate)
      return null;
    if (mappings.length >= 65536)
      throw Exception("KeysTable.getKeyIndex: too many keys");
    mappings.add(KeyMapping(mappings.length, key));
    count++;
    return mappings.length - 1;
  }

  void visitKey(String key) {
    getKeyIndex(key, allowCreate: true);
  }
}

File: python_lib/structs/KeysTable.py
from FileWrapper import BinaryReader, BinaryWriter
from structs.KeyMapping import KeyMapping


class KeysTable:
	count: int
	mappings: list[KeyMapping]

	def __init__(self):
		self.count = 0
		self.mappings = []

	@staticmethod
	def readBytes(bytes: BinaryReader):
		keysTable = KeysTable()
		keysTable.count = bytes.readUint16()
		keysTable.mappings = [KeyMapping.readBytes(bytes) for _ in range(keysTable.count)]
		return keysTable

	def writeBytes(self, bytes: BinaryWriter):
		bytes.writeUint16(self.count)
		for mapping in self.mappings:
			mapping.writeBytes(bytes)

	@property
	def size(self) -> int:
		size = 2
		for mapping in self.mappings:
			size += mapping.size
		return size

	def lookupIndex(self, index: int) -> str:
		return self.mappings[index].keyName

	def getKeyIndex(self, key: str, allowCreate: bool = False) -> int|None:
		for i in range(len(self.mappings)):
			if self.mappings[i].keyName == key:
				return i
		if not allowCreate:
			return None
		if len(self.mappings) >= 65536:
			raise Exception("KeysTable.getKeyIndex: too many keys")
		self.mappings.append(KeyMapping(len(self.mappings), key))
		self.count += 1
		return len(self.mappings) - 1

	def visitKey(self, key: str):
		self.getKeyIndex(key, allowCreate=True)
